Introduce tizen_interop package

WonyoungChoi opened the pull request:
context: https://github.com/flutter-tizen/flutter-tizen/issues/129

This `tizen_interop` package provides Dart bindings and utility functions for using Tizen native APIs.

### Example for using Interop
```dart
import 'dart:ffi';
import 'package:ffi/ffi.dart';
import 'package:tizen_interop/4.0/wearable/tizen.dart';

String? getApplicationName() {
    String? appName;
    using((Arena arena) {
      Pointer<Pointer<Int8>> namePtr = arena();
      tizen.app_get_name(namePtr);
      appName = namePtr.value.toDartString();
      malloc.free(namePtr.value);
    });
    return appName;
}
```

swift-kim commented on LICENSE:
```diff
@@ -0,0 +1,25 @@
+      copyright notice, this list of conditions and the following
+      disclaimer in the documentation and/or other materials provided
+      with the distribution.
+    * Neither the names of the copyright holders nor the names of the
```
```suggestion
    * Neither the name of the copyright holder nor the names of the
```

swift-kim commented on README.md:
```diff
@@ -0,0 +1,22 @@
+
+String? getApplicationName() {
+    String? appName;
+    using((Arena arena) {
```
Double indentations.

swift-kim commented on lib/4.0/wearable/tizen.dart:
```diff
@@ -0,0 +1,42 @@
+      throw ArgumentError(
+          'Failed to lookup symbol (undefined symbol: $symbolName');
+    }
+    var library = _libraryCache[libraryName];
+    if (library == null) {
+      _libraryCache[libraryName] = DynamicLibrary.open(libraryName);
+    }
```
```suggestion
    _libraryCache[libraryName] ??= DynamicLibrary.open(libraryName);
```

swift-kim commented on packages/symgen/lib/src/config.dart:
```diff
@@ -0,0 +1,91 @@
+    }
+    final file = LocalFileSystem().file(path);
+    if (!file.existsSync()) {
+      throw FormatException('$path is not found.');
```
`FileSystemException`?

swift-kim commented on packages/symgen/lib/src/config.dart:
```diff
@@ -0,0 +1,91 @@
+    dynamic value = this[key];
+    if (value == null) {
+      if (defaultValue == null && mandatory) {
+        throw FormatException('$key not found');
```
`StateError`?

swift-kim commented on packages/symgen/lib/src/generator.dart:
```diff
@@ -0,0 +1,58 @@
+    var fs = const LocalFileSystem();
+    final libraryPath = fs.directory(_config.libraryPath);
+    if (!libraryPath.existsSync()) {
+      throw FormatException('$libraryPath is not found.');
```
Same here!

swift-kim commented on packages/symgen/lib/src/generator.dart:
```diff
@@ -0,0 +1,58 @@
+${_config.preamble}
+// AUTO GENERATED FILE, DO NOT EDIT.
+//
+// Generated by `package:symgen`.'
```
```suggestion
// Generated by `package:symgen`.
```

swift-kim commented on pubspec.yaml:
```diff
@@ -0,0 +1,17 @@
+name: tizen_interop
+description: Provides Dart bindings and utility functions for Tizen native APIs.
+version: 0.0.1
+homepage: https://developer.tizen.org
```
```suggestion
homepage: https://github.com/flutter-tizen/tizen_interop
```
This URL should be more useful for users, if you're going to upload this package to pub.dev.

swift-kim commented on packages/symgen/lib/src/config.dart:
```diff
@@ -0,0 +1,91 @@
+    config._targetLibraries =
+        List.unmodifiable(map.getValue('target-libraries'));
+
+    config._whiteList.addAll(_readWhiteListFile(
+      map.getValue('white-list', mandatory: false),
```
```suggestion
    config._allowlist.addAll(_readAllowlistFile(
      map.getValue('allowlist', mandatory: false),
```

swift-kim commented:
Some warnings were generated while running `build.sh`.
```
Running in Directory: '/home/swift/Git/tizen_interop'
[WARNING]: Prefer adding Key '[description]' to your config.
Input Headers: [configs/4.0/wearable/entrypoints.h]
[SEVERE] : Header configs/4.0/wearable/entrypoints.h: Total errors/warnings: 1.
[SEVERE] :     ./rootstraps/wearable-4.0-device.core/usr/include/pthread.h:743:12: warning: declaration of built-in function '__sigsetjmp' requires the declaration of the 'jmp_buf' type, commonly provided in the header <setjmp.h>. [Semantic Issue]
[WARNING]: Removed All Struct Members from camera_preview_data_s(camera_preview_data_s), struct member has an unsupported type.
[WARNING]: Removed All Struct Members from calendar_time_s(calendar_time_s), struct member has an unsupported type.
[WARNING]: Skipped Function 'calendar_filter_add_caltime', Incomplete struct pass/return by value not supported.
[WARNING]: Skipped Function 'calendar_record_set_caltime', Incomplete struct pass/return by value not supported.
[WARNING]: Generated declaration '_watch_time_s' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_activity_handle_s' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_activity_data_s' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_gesture_handle_s' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_gesture_data_s' start's with '_' and therefore will be private.
...
...
```
Is this normal?

WonyoungChoi commented:
@swift-kim I think `underscore` warnings are not issues because they are not public apis. But others should be checked more.

swift-kim commented on packages/symgen/lib/src/executable.dart:
```diff
@@ -0,0 +1,63 @@
+      'config',
+      help: 'Path to Yaml file containing configurations',
+      mandatory: true,
+    );
```
How about adding an option that takes a path to the allowlist file, or adding an entry
```yaml
allowlist: ${TIZEN40_WEARABLE_ALLOWLIST_PATH}
```
to `configs/4.0/wearable/symgen.yaml` file so that generating the bindings can be easily automated?

WonyoungChoi replied:
👍 good idea.

WonyoungChoi replied:
@swift-kim I added options to symgen to override `output`, `allowlist`, `library-path`.

swift-kim replied:
Thanks!

WonyoungChoi commented:
> Some warnings were generated while running `build.sh`.
> 
> ```
> Running in Directory: '/home/swift/Git/tizen_interop'
> [WARNING]: Prefer adding Key '[description]' to your config.
> Input Headers: [configs/4.0/wearable/entrypoints.h]
> [SEVERE] : Header configs/4.0/wearable/entrypoints.h: Total errors/warnings: 1.
> [SEVERE] :     ./rootstraps/wearable-4.0-device.core/usr/include/pthread.h:743:12: warning: declaration of built-in function '__sigsetjmp' requires the declaration of the 'jmp_buf' type, commonly provided in the header <setjmp.h>. [Semantic Issue]
> [WARNING]: Removed All Struct Members from camera_preview_data_s(camera_preview_data_s), struct member has an unsupported type.
> [WARNING]: Removed All Struct Members from calendar_time_s(calendar_time_s), struct member has an unsupported type.
> [WARNING]: Skipped Function 'calendar_filter_add_caltime', Incomplete struct pass/return by value not supported.
> [WARNING]: Skipped Function 'calendar_record_set_caltime', Incomplete struct pass/return by value not supported.
> [WARNING]: Generated declaration '_watch_time_s' start's with '_' and therefore will be private.
> [WARNING]: Generated declaration '_activity_handle_s' start's with '_' and therefore will be private.
> [WARNING]: Generated declaration '_activity_data_s' start's with '_' and therefore will be private.
> [WARNING]: Generated declaration '_gesture_handle_s' start's with '_' and therefore will be private.
> [WARNING]: Generated declaration '_gesture_data_s' start's with '_' and therefore will be private.
> ...
> ...
> ```
> 
> Is this normal?

if the `union` is used in a struct, `ffigen 3.0.0` can't generate the Dart code. but it will be resolved in next release version of `ffigen`. 
ref: https://github.com/dart-lang/ffigen/issues/214

bwikbs commented:
Wow! Found this just now! I am very interested in this subject! :+1: 

Although it's very early stage, but.. I've question. 
This is for tizen dart api  right? If so, do you have a idea how building dart interface too?
When we reviewed FFI before, we gave up because It is not easy to get users to use the output of FFI.  And we didn't know how to make tizen dart api interface either.
So, if you have any rough ideas for next step, I'd love to know that!

WonyoungChoi commented:
@bwikbs Thank you for your interest. 😃 
As you mentioned, the output of `ffigen` is not easy for end-user to use. In particular, pointer handling is not kind because there is no `finalizer` in dart-lang... 
The project is designed to open up a minimal way to use the Tizen Native API, although it is inconvenient to use. It can also help create a new dart interfaces. 
And.. I plan to add more utilities to make it easier to use the ffigen's output..

bwikbs commented:
@WonyoungChoi Thanks for letting me know!

>   I plan to add more utilities to make it easier to use the ffigen's output..

Can you share roughly what you are planning? Maybe I'll get some inspiration from your plans :smile:

bwikbs commented:
Oh, one more thing.. :smile: (just curious,after it came out..)
What do you think of the challenging task to introduce ``rest api`` or ``grpc``? Or build another meta-data for this propose.

swift-kim commented:
@WonyoungChoi Have you taken a look at the issue regarding `__sigsetjmp` in the above?

Otherwise the change looks good to me.

WonyoungChoi commented:
> @WonyoungChoi Thanks for letting me know!
> 
> > I plan to add more utilities to make it easier to use the ffigen's output..
> 
> Can you share roughly what you are planning? Maybe I'll get some inspiration from your plans 😄

Well.. first of all, I'm looking for a more comfortable way to use the c style string. For now, the following procedure is required to receive string from the native function.
```dart
Pointer<Pointer<Int8>> ppString = malloc();
final ret = tizen.some_native_function(ppString);
final dartStr = ppString.value.toDartString();
malloc.free(ppString.value);
malloc.free(ppString);
```

WonyoungChoi commented:
> @WonyoungChoi Have you taken a look at the issue regarding `__sigsetjmp` in the above?
> 
> Otherwise the change looks good to me.

Could you explain it in more detail? I can't find a issue related `__sigsetjump()`.

WonyoungChoi commented:
> Oh, one more thing.. 😄 (just curious,after it came out..)
> What do you think of the challenging task to introduce `rest api` or `grpc`? Or build another meta-data for this propose.

Sorry, I can't understand the exact point of the part you mentioned. But I found https://github.com/grpc/grpc-dart..

swift-kim commented:
> Could you explain it in more detail? I can't find a issue related `__sigsetjmp()`.

@WonyoungChoi You can just search for the text in this page.

bbrto21 commented:
Q. Does the tool create the files below?  Why do we need to include these?

- generated_bindings.dart
- generated_symbols.dart

WonyoungChoi commented:
> > Could you explain it in more detail? I can't find a issue related `__sigsetjmp()`.
> 
> @WonyoungChoi You can just search for the text in this page.

Oh. I missed that :( Because there is no `__sigsetjmp()` errors in my log after `build.sh`. Could you try again?

WonyoungChoi commented:
> Q. Does the tool create the files below? Why do we need to include these?
> 
> * generated_bindings.dart
> * generated_symbols.dart

Yes these files are generated by tools, but they should be included into the package. Of course, they can be generated every time  the package is published.  
I think it's easier to use these files being managed in git. the `build.sh` only needs to be used to update these files.

bwikbs commented:
> Sorry, I can't understand the exact point of the part you mentioned. But I found https://github.com/grpc/grpc-dart..

It's not a big deal. I was asking if you've ever thought about how to build a tizen interface with other general-purpose technologies and binding dart from that. (In fact, there is a rumor from vd  and I was asking you heard or knew anything about that :smile: )

swift-kim commented:
@WonyoungChoi The error is still reproducible, but I could suppress the error by adding `-Wno-incomplete-setjmp-declaration` to `ffigen.yaml`'s `compiler-opts:`.

WonyoungChoi commented:
> @WonyoungChoi The error is still reproducible, but I could suppress the error by adding `-Wno-incomplete-setjmp-declaration` to `ffigen.yaml`'s `compiler-opts:`.

@swift-kim  Thank you for finding useful option! It seems to be related to the "clang/llvm" environment. anyway I don't think it is not problem because the `pthread.h` is not a target of ffigen. I will add that option to ffigen.yaml file also.
